lms/db/_session_extension.py

```diff
@@ -0,0 +1,29 @@
+class SessionExtension:
+    """
+    A self returning descriptor to adding functionality to a session.
+
+    To use this, sub-class it and then add your object to the session:
+
+    class CustomSession(Session):
+        my_extension = MyExtension()
+
+    The `name` and `session` attributes will be available to your instance.
+    """
+
+    session = None
+    name = None
+
+    def __set_name__(self, owner, name):
+        # Store the name for more informative errors
+        self.name = name
+
+    def __get__(self, instance, owner):
+        if not instance:
+            raise TypeError(
+                f"You can only call '{self.name}' on instances not the bare "
+                f"class: '{owner.__name__}'"
+            )
+
+        # Store the session and return ourselves
+        self.session = instance
+        return self
```

Comment:
This indirection layer is here 100% to facilitate having the bulk functionality bolt onto the session object rather than be something directly provided by the session object.

If we don't care about that and are happy to put the bulk action functionality in the session directly there's no point in this.

Another way to achieve the same thing would be:

```python
class BulkAction:
    def __init__(self, session):
        self.session = session

class CustomSession(Session):    
    @cached_property
    def bulk(self):
        return BulkAction(self)
```

Which achieves much the same thing but requires more co-operation from the session object.